Documentation for test suite (#298)
CONP-PCNO/conp-dataset: tests/README.md

        ┌──────────────────◄──────────────────────┘
        │
        ▼
        │                                                    worker x2
┌───────┴───────┐         ┌──────────────────┐         ╔═════════════════╗         ┌───────────────────┐
│ Create Tests  ├────►────┤    Split Tests   ├────►────╢       Test      ╟────►────┤   Tests Results   │
╞═══════════════╡         ╞══════════════════╡         ╠═════════════════╣         ╞═══════════════════╡
│ Generate test │         │ Distribute tests │         ║ datalad_install ║         │ Show on dashboard │
│   files from  │         │  equally amongs  │         ║     (Setup)     ║         ├───────────────────┤
│    Template   │         │     workers      │         ╟─────────────────╢         │ Save to artifacts │
└───────────────┘         └──────────────────┘         ║    has_readme   ║         └─────────┬─────────┘
                                                       ╟─────────────────╢                   │
                                                       ║ has_valid_dats  ║                   │
                                                       ╟─────────────────╢                   │

>                                                        ║   datalas_get   ║                   ▼

Reviewer: * `datalas` -> `datalad`.

**Test job:**

- Parallelism on 2 workers.
- Execute the dataset template test suite.

After test job: (WIP)<br/>
The test results are parsed and save to CircleCI artifacts.
#TODO Send them to a monitoring GitHub Repository

## Life of a Dataset Test

This section describes the different steps a dataset test will perform to validate the proper functioning of datasets.

### Test creation

At the start of the test job, the `template.py` file is used to generate a test file for each dataset.
Changes made to a dataset should not influence the behavior of other datasets while changes to the test suite potentially impacts every dataset.
To solve this, the test suite will start by retrieving all files modified during a pull request.
Then, it will only test on the minimal set of datasets.
This strategy aims at saving computing resources as well as time.

To select whether a dataset should be tested or not, we follow this heuristic:

1. File modified is **not** part of a dataset:
   - is part of a whitelist (see below), then **ignore**;
   - otherwise, **full rerun** of the test suite.

> 2. File is part of a dataset, then **partial rerun** test for this dataset.

Reviewer: You should explain that when we're not in a PR then all datasets are tested.

Author: Brought the change. Let me know if I should add further emphasis on this point

#### Datalad install **(Setup)**

This steps runs before every test case mentioned below.
To prevent concurrent execution of the command to cause issue, test case for a dataset are queued when the dataset is already being installed.

#### Has `README.md`

Every dataset is required to have a `README.md` file to describe its content.
This test validates that this file exists.

#### Has a valid `DATS.json` file

Every dataset must have a `DATS.json` that follows that [DATS model convention](https://datatagsuite.github.io/docs/html/dats.html).
This test makes sure that the `DATS.json` file exist and that its content is conformed to the DATS Model.

#### Datalad get

The ability to download a dataset is a critical component for its proper function.
This test goes through the process of downloading files from a dataset to valid its validity.
That is, the test case executes those steps:

1. Authenticate the dataset (see [Authenticated Dataset](#authenticated-dataset) section).
1. (**Travis specific**) Remove files using FTP due to issue on travis; more detail [here](https://blog.travis-ci.com/2018-07-23-the-tale-of-ftp-at-travis-ci).

> 1. Select a k files, with smallest size, from a sample of n files. This aims at saving computing resources (see [Timeout](#timeout) and [Size of Annexed File](#size-of-annexed-file) sections).

Author: Would that sentence be easier to understand: "From a sample of n files, the k smallest files are selected."

> 1. Select a k files, with smallest size, from a sample of n files. This aims at saving computing resources (see [Timeout](#timeout) and [Size of Annexed File](#size-of-annexed-file) sections).

Reviewer 2: I think so. Thank you!

### Secrets creation

To create the secrets in CircleCI you will need to generate a standardized \${PROJECT_NAME} as done below:

```python
from tests.functions import project_name2env
project_name2env("projects/dataset_name".split("/")[-1])
# DATASET_NAME
```

#### Loris

1. Create a LORIS account to be used on CircleCI.
1. Assisted by the CircleCI administrator create the following secrets:
   - `${PROJECT_NAME}_USERNAME`
   - `${PROJECT_NAME}_PASSWORD`
   - `${PROJECT_NAME}_LORIS_API`; e.g. `https://example.loris.ca/api/v0.0.0`

#### Zenodo

1.  [Create a new Zenodo token](https://zenodo.org/account/settings/applications/tokens/new/)
1.  Assisted by the CircleCI administrator, create a secret called `${PROJECT_NAME}_ZENODO_TOKEN`.

### Limitations of secrets

A major limitation of secrets is that authenticated datasets cannot be tested on pull requests due to privacy issues.
Indeed, malicious users could easily retrieve secrets by making a pull requests.

> To avoid this problematic, the user making a pull request on an authenticated dataset can set up secrets for the dataset in its CircleCI account.

Reviewer: problematic -> problem.

#TODO

> Otherwise, the pull requests for authenticated datasets should be merged into a devel branch to assure their proper functioning.

Author: @glatard do you have updates on creating a development branch?
I will match the names afterwards.

Then, when everything is working, it can be merged into the master branch.

## Implementation Keypoints

### Flaky test

The current test suite cover a broad range of components for dataset to work properly. Unfortunately, there are components that are hard to test in real environment such as downloading files from live servers.
<br/>
For this reason, we opted to allow test to be flaky.
That is, when a test fails it will rerun up to 3 times.
There is a 5 seconds delay between rerun of failures.
Furthermore, to avoid Datalad command to conflict during a dataset installation, an instruction to install a dataset will be queued if this dataset is already installing.

### Empty dataset

> When a dataset has no files contained in its annex, the test suite assume the dataset works properly and pass the test.

Reviewer 2: ```suggestion
When a dataset has no file contained in its annex, the test suite assumes the dataset works properly and passes the test.
```

### Timeout

Since the test suite involves downloading data from external servers, the runtime of the test suite can vary considerably when there is network issues.
To mitigate bad connection to server, the test suite imposes a 10-minute timeout for each test as well as a 2-minute timeout specific to each dataset download.

### Size of annexed file

Dataset files vary in size.

> This can be and issue when testing the download of a dataset as downloading a large file might result in timeout and test failure.

Reviewer 2: ```suggestion
This can be an issue when testing the download of a dataset as downloading a large file might result in a timeout and test failure.
```

> To prevent this problem to happen, the test suite sorts the files based on their size; if a file size cannot be retrieved it sets it to infinity.

Reviewer 2: ```suggestion
To prevent this problem, the test suite sorts the files based on their size; if a file size cannot be retrieved it sets it to infinity.
```

Then, it downloads the files in increasing order of their size.

> This as the potential to speedup the test suite considerably.

Reviewer 2: ```suggestion
This has the potential to speedup the test suite considerably.
```
